src/platforms/hosted/jlink.c

```diff
@@ -277,6 +234,313 @@ static bool jlink_claim_interface(void)
+			const bool is_bmda_supported = i == JLINK_INTERFACE_SWD || i == JLINK_INTERFACE_JTAG;
+
+			DEBUG_INFO("\t%zu: %s%c %s\n", i, jlink_interface_to_string(i), is_current ? '*' : ' ',
+				is_bmda_supported ? "" : "(Not supported)");
+		}
+	}
+
+	return true;
+}
+
+static bool jlink_get_interface_frequency(const uint8_t interface)
+{
+	if (!(jlink.capabilities & JLINK_CAPABILITY_INTERFACE_FREQUENCY)) {
+		DEBUG_WARN("J-Link does not support interface frequency commands\n");
+		return false;
+	}
+
+	/* If the requested interface is invalid, bail out */
+	if (!jlink_interface_available(interface))
+		return false;
+
+	/* If the base frequency is non-zero, we've already read the frequency info for the requested interface and can skip the query */
+	if (jlink.interface_frequency[interface].base != 0U)
+		return true;
+
+	/* Get the selected interface */
+	const uint8_t selected_interface = jlink_selected_interface();
+
+	if (selected_interface != interface) {
+		/* If the selected interface doesn't match the requested interface, select it, let's hope this doesn't mess something up elsewhere */
+		DEBUG_WARN("Trying to get frequency for interface %s but it is not selected, selecting it\n",
+			jlink_interface_to_string(interface));
+
+		/* select the requested interface */
+		if (!jlink_select_interface(interface))
+			return false;
+	}
+
+	/* Get the frequency info for the selected interface */
+	uint8_t buffer[6U];
+	if (!jlink_simple_query(JLINK_CMD_INTERFACE_GET_BASE_FREQUENCY, buffer, sizeof(buffer)))
+		return false;
+
+	struct jlink_interface_frequency *const interface_frequency = &jlink.interface_frequency[interface];
+
+	interface_frequency->base = read_le4(buffer, JLINK_INTERFACE_BASE_FREQUENCY_OFFSET);
+	interface_frequency->min_divisor = read_le2(buffer, JLINK_INTERFACE_MIN_DIV_OFFSET);
+
+	/* This is an assumption, if the J-Link was configured before we started, this may not be true, but we have no way to know */
+	interface_frequency->current_divisor = interface_frequency->min_divisor;
+
+	DEBUG_INFO("%s interface frequency:\n\tBase frequency: %uHz\n\tMinimum divisor: %u\n",
+		jlink_interface_to_string(interface), interface_frequency->base, interface_frequency->min_divisor);
+
+	// Restoring the selected interface seemed to cause issues sometimes (unsure if culprit)
+	// so we don't do it for now, it didn't seem to be necessary, saving in case it proves to be necessary later
+	// if (selected_interface != interface) {
+	// 	/* If the selected interface didn't match the requested interface, restore the selected interface */
+	// 	DEBUG_WARN("Restoring selected interface to %s\n", jlink_interface_to_string(selected_interface));
+
+	// 	/* select the requested interface, we don't consider this a critical failure, as we got what we wanted */
+	// 	if (!jlink_select_interface(selected_interface))
+	// 		DEBUG_ERROR("Failed to restore selected interface to %s\n", jlink_interface_to_string(selected_interface));
+	// }
```

Comment:
Rather than committing this as commented out code, it would be better to instead commit this in an `#if 0` block if it's going to be included in the PR